Implemented single link instead of multiple links to read reviews

muskangupta-iitr opened the pull request:
<!--
    Thanks for opening a pull request (PR). Please read through
    these instructions to help us review and merge your change quicker.

    Replace ISSUE_NUMBER with the number of the issue related to what
    you're fixing followed by a description of your change. For example:

    Fixes #3588

    This patch adds a margin on the logout button to correct the layout.
-->

This patch implements the approach of the single link instead of multi links as mentioned in the patch #9254 .
Fixes #8860 

@muffinresearch @willdurand @bobsilverberg Please review the code!
<!--
    Read through this checklist to make sure your patch is ready for review.

    - Add a PR title that summarizes your patch
    - Make sure this PR relates to an existing open issue and there are no existing PRs open for the same issue.
    - Add tests to cover the changes added in this PR.
    - Check that the change works locally.
    - Add before and after screenshots (Only for changes that impact the UI).

    Thanks for your contribution!
-->

willdurand commented:
How is this related to https://github.com/mozilla/addons-frontend/issues/8827?

willdurand commented:
I am going to close this patch for now, even if that's just a draft. Please focus on:

- https://github.com/mozilla/addons-frontend/pull/9253
- https://github.com/mozilla/addons-frontend/pull/9254

muskangupta-iitr commented:
@willdurand Sorry, by mistakenly I tagged different issue on this, I apologize for that. I updated the above comment. Really sorry for such unprofessional behavior. @muffinresearch Only told me to create this patch to try different approach for implementing single link instead of multiple links. Here is the [comment](https://github.com/mozilla/addons-frontend/pull/9254#issuecomment-604064434). I already successfully implemented the first approach as in the [comment](https://github.com/mozilla/addons-frontend/pull/9254#issuecomment-603899962)

muskangupta-iitr commented:
@willdurand Can you review and open this patch please as I have to proceed further with these issues?

muskangupta-iitr commented:
@muffinresearch @willdurand Should I create a new path for this?

willdurand commented:
Oh, I see. Sorry, then let's reopen this PR and close thd other one, ok?

muskangupta-iitr commented:
I had created two PRs according to two different approaches as discussed before. You can review the two PRs so that whichever is found suitable by you, could be further proceeded with while closing the other.

muskangupta-iitr commented:
@bobsilverberg I tried the approach of making row into individual grid but in that case the width of each bar will be differently calculated and independent to each other that's why I breaked them into manual columns so that all the bar width will be calculated according to max-width of the column.

bobsilverberg commented:
> @bobsilverberg I tried the approach of making row into individual grid but in that case the width of each bar will be differently calculated and independent to each other that's why I breaked them into manual columns so that all the bar width will be calculated according to max-width of the column.

Ah, I thought my method was working:

<img width="475" alt="Screenshot 2020-03-31 17 18 41" src="https://user-images.githubusercontent.com/142755/78076387-53f22480-7374-11ea-9b16-1295ba85cd75.png">

but then realized (after you pointed it out) that if the review count was a different number of digits for different rows it doesn't work. Hmm, let me take another look.

muskangupta-iitr commented:
@bobsilverberg Yes, I am also facing the same problem when the width of the third column will be different. Then, I switched to table for solving this.

bobsilverberg commented:
@muskangupta-iitr But it looks like the same problem exists with your table approach:

<img width="479" alt="Screenshot 2020-03-31 18 26 50" src="https://user-images.githubusercontent.com/142755/78080931-4ab98580-737d-11ea-9950-4f7e73d6eb6b.png">

Are you not seeing this issue with your patch?

muskangupta-iitr commented:
@bobsilverberg Here is the screenshot from my patch:

![Screenshot from 2020-04-01 04-26-53](https://user-images.githubusercontent.com/50856599/78082606-14a5f100-73d1-11ea-87bd-d83be39e0630.png)

bobsilverberg commented on src/amo/components/RatingsByStar/index.js:
```diff
@@ -86,85 +90,104 @@ export class RatingsByStarBase extends React.Component<InternalProps> {
+    const { addon, location, history, clientApp, siteLang } = this.props;
+    history.push({
+      pathname: `/${siteLang}/${clientApp}/addon/${addon.slug}/reviews/`,
+      search: `?score=${star}&src=${location.query.src}`,
```
If `src` does not exist on the URL, then this yields a link with `src=undefined` in it, which I don't think we want.

muskangupta-iitr replied:
Yes, we should add a conditional check for `src`. I will update that.

bobsilverberg commented on src/amo/components/RatingsByStar/index.js:
```diff
@@ -86,85 +90,104 @@ export class RatingsByStarBase extends React.Component<InternalProps> {
-      '2': i18n.gettext('Read all two-star reviews'),
-      '1': i18n.gettext('Read all one-star reviews'),
-      /* eslint-enable quote-props */
+    const self = this;
```
Why is this needed?

muskangupta-iitr replied:
Beacuse I have to use `this` in `createLink` function. As, `this` will be undefined in the scope of that function.

bobsilverberg commented on src/amo/components/RatingsByStar/index.js:
```diff
@@ -86,85 +90,104 @@ export class RatingsByStarBase extends React.Component<InternalProps> {
+
+                return (
+                  <tr
+                    title={getLinkTitle(star, starCount) || ''}
```
`getLinkTitle` returns `''` as the `default`, so why do you need the `|| ''` here?

muskangupta-iitr replied:
Yeah, I added this as extra check type but yes can be removed. I will updated that.

bobsilverberg commented on src/amo/components/RatingsByStar/index.js:
```diff
@@ -29,11 +30,14 @@ type Props = {|
+  history: ReactRouterHistoryType,
   i18n: I18nType,
   location: ReactRouterLocationType,
+  siteLang: ?string,
```
I don't think we need the `?` here. `siteLang` should always exist.

muskangupta-iitr replied:
Ohk! I will change this in the next update.

bobsilverberg commented on src/amo/components/RatingsByStar/index.js:
```diff
@@ -86,85 +90,104 @@ export class RatingsByStarBase extends React.Component<InternalProps> {
                   {loading ? (
-                    <div className="RatingsByStar-bar RatingsByStar-barFrame" />
+                    <tr>
+                      <th className="RatingsByStar-star">
```
Why are you using `th` rather than `td`? As these are data rows, and now a header row, `td` seems more appropriate to me.

muskangupta-iitr replied:
Yes, I just tried the table approach with `th` tag, but `td` will be more appropriate. I will update that too.

bobsilverberg commented on src/amo/components/RatingsByStar/index.js:
```diff
@@ -86,85 +90,104 @@ export class RatingsByStarBase extends React.Component<InternalProps> {
-                <div className="RatingsByStar-barContainer">
+              return (
+                <React.Fragment key={star}>
                   {loading ? (
```
I wonder if it would be a bit more efficient and cleaner if you were to produce the contents of the `<tr>` via a function, which returns either the loading version or the non-loading version, and then here either wrap that in `createLink` or not, depending on `loading`? This would allow you to avoid all the duplicate code for each `th` tag.

muskangupta-iitr commented:
@bobsilverberg Done! I have updated the PR with the changes discussed above :+1:

bobsilverberg commented on src/amo/components/RatingsByStar/index.js:
```diff
@@ -86,85 +90,108 @@ export class RatingsByStarBase extends React.Component<InternalProps> {
+    const { addon, location, history, clientApp, siteLang } = this.props;
+    history.push({
+      pathname: `/${siteLang}/${clientApp}/addon/${addon.slug}/reviews/`,
+      search: `?score=${star}&${
```
I think you need to put the `&` inside the `{ }` for `location.query.src` because the way the code is now, without a `src` you still get a trailing `&` on the URL.

muskangupta-iitr replied:
Actually while writing tests for it I found a better approach for doing this. It goes like this:
```
    history.push(
      `/${siteLang}/${clientApp}/${reviewListURL({
        addonSlug: addon.slug,
        score: star,
        src: location.query.src,
      })}`,
    );
```
which will use the general function for creating url.

muskangupta-iitr commented:
@bobsilverberg Umm, no actually I was trying few other approaches that could be used and was confused between which could be better so forgot to comment. I have already addressed the comment in my last commit with a different approach and would request you to review it and give suggestions so that we could implement whichever is best. :)

muskangupta-iitr commented:
@bobsilverberg I have added the tests and changed the `history.push` path. So, I think this PR is ready for review?

bobsilverberg commented on src/amo/components/RatingsByStar/index.js:
```diff
@@ -86,85 +91,113 @@ export class RatingsByStarBase extends React.Component<InternalProps> {
+    invariant(addon, 'addon was unexpectedly empty');
+
+    history.push(
+      `/${siteLang}/${clientApp}${reviewListURL({
```
This is better, nice job!

muskangupta-iitr replied:
@bobsilverberg Thanks!

muskangupta-iitr commented:
> Thanks for the updates @muskangupta-iitr. I added one comment, but also I don't think you addressed my comment at [#9295 (comment)](https://github.com/mozilla/addons-frontend/pull/9295#discussion_r401624501) yet.

@bobsilverberg Yes, the above changes looks really good. Thanks:). Should I replace mine with yours?

bobsilverberg commented:
> > Thanks for the updates @muskangupta-iitr. I added one comment, but also I don't think you addressed my comment at [#9295 (comment)](https://github.com/mozilla/addons-frontend/pull/9295#discussion_r401624501) yet.
> 
> @bobsilverberg Yes, the above suggestion looks really better, could not think of this approach. Its actually a really good approach, reducing code redundancy as well as working really efficiently in my local system. Thanks for the suggestion. May I proceed with replacing my code with your suggestion?

Yes, please go ahead and use any or all of that suggestion in your patch.

bobsilverberg commented on tests/unit/amo/components/TestRatingsByStar.js:
```diff
@@ -148,30 +155,59 @@ describe(__filename, () => {
+
+      row.simulate('click', createFakeEvent({}));
+
+      sinon.assert.calledWithExactly(
```
We tend to use just `sinon.assert.calledWith` as opposed to `sinon.assert.calledWithExactly`. Is there a reason you chose to use `calledWithExactly` here?

muskangupta-iitr replied:
Like there is no specific reason for using it, but I saw some test in which `calledWithExactly` is used, in `TestSearchFilter`. i will update that.

bobsilverberg commented on tests/unit/amo/components/TestRatingsByStar.js:
```diff
@@ -148,30 +155,59 @@ describe(__filename, () => {
+    const tableRow = root.find('.RatingByStar-table-row');
+
+    function validateLink(row, score, expectedTitle) {
+      const clientApp = CLIENT_APP_ANDROID;
```
This test is using values that will be present because they are the default values, but we really shouldn't rely on those in a test. If we're going to have an assertion that includes `clientApp` and/or `lang`, then we should explicitly set those in the test, and the assertion should check for the values we set. You can see an example of how this can be done in the test for ['renders an opensearch link for Firefox' in TestSearchForm.js](https://github.com/mozilla/addons-frontend/blob/master/tests/unit/amo/components/TestSearchForm.js#L179-L190).

muskangupta-iitr replied:
Thanks @bobsilverberg :) Actually, I was looking this type of approach only but could not able to find it. I will update this.

muskangupta-iitr replied:
@bobsilverberg I tried this approach but getting an error.
This is the code I implemented:
```
  it('adds a `src` query parameter to the review links when available in the location', () => {
    const src = 'some-src';
    const grouping = {
      5: 964,
      4: 821,
      3: 543,
      2: 22,
      1: 0,
    };
    const addon = addonForGrouping(grouping);
    const clientApp = CLIENT_APP_FIREFOX;
    const lang = 'fr';

    const customStore = dispatchClientMetadata({ clientApp, lang }).store;
    customStore.dispatch(setGroupedRatings({ addonId: addon.id, grouping }));

    const root = render({
      addon,
      location: createFakeLocation({ query: { src } }),
    });
    const tableRow = root.find('.RatingByStar-table-row');

    function validateLink(row, score) {
      row.simulate('click', createFakeEvent());

      sinon.assert.calledWith(
        fakeHistory.push,
        `/${lang}/${clientApp}${reviewListURL({
          addonSlug: addon.slug,
          score,
          src,
        })}`,
      );
    }

    validateLink(tableRow.at(0), '5');
    validateLink(tableRow.at(1), '4');
    validateLink(tableRow.at(2), '3');
    validateLink(tableRow.at(3), '2');
    validateLink(tableRow.at(4), '1');
  });
```

Here is the error:
```
 AssertError: expected stub to be called with arguments

      241 |       row.simulate('click', createFakeEvent());
      242 | 
    > 243 |       sinon.assert.calledWith(
          |                    ^
      244 |         fakeHistory.push,
      245 |         `/${lang}/${clientApp}${reviewListURL({
      246 |           addonSlug: addon.slug,

      at Object.fail (node_modules/sinon/lib/sinon/assert.js:107:21)
      at failAssertion (node_modules/sinon/lib/sinon/assert.js:66:16)
      at Object.calledWith (node_modules/sinon/lib/sinon/assert.js:92:13)
      at validateLink (tests/unit/amo/components/TestRatingsByStar.js:243:20)
      at Object.<anonymous> (tests/unit/amo/components/TestRatingsByStar.js:253:5)

```

muskangupta-iitr replied:
I did a really stupid mistake :sweat_smile: I solved it. I forgot to pass history and store to render method. Thanks :)

bobsilverberg commented on tests/unit/amo/components/TestRatingsByStar.js:
```diff
@@ -148,30 +155,59 @@ describe(__filename, () => {
+
+      expect(row).toHaveProp('onClick');
+
+      row.simulate('click', createFakeEvent({}));
```
You don't need the `{}` here, it can just be `createFakeEvent()`.

muskangupta-iitr replied:
Yes,I will update this.

bobsilverberg commented on tests/unit/amo/components/TestRatingsByStar.js:
```diff
@@ -148,30 +155,59 @@ describe(__filename, () => {
     store.dispatch(setGroupedRatings({ addonId: addon.id, grouping }));
     const root = render({ addon });
-    const counts = root.find('.RatingsByStar-star').find(Link);
-    const bars = root.find('.RatingsByStar-barContainer').find(Link);
```
I think the description of this test should be updated, as I think it is no longer accurate. For example, it's not asserting anything about bars.

muskangupta-iitr replied:
@bobsilverberg Should I change the description or add test for bars ?

muskangupta-iitr replied:
I my opinion because we have separate test case for bars, we can change the description of the test. So, after merging star count test description will be `renders star labels, star counts and links`.

muskangupta-iitr replied:
@bobsilverberg Thanks!

bobsilverberg commented on tests/unit/amo/components/TestRatingsByStar.js:
```diff
@@ -185,22 +221,14 @@ describe(__filename, () => {
+    const tableRow = root.find('.RatingByStar-table-row');
+    const counts = tableRow.find('.RatingsByStar-count');
+
+    expect(counts.at(0).children()).toHaveText(grouping['5'].toString());
```
I wonder if you can just add these assertions into the test you changed above, as opposed to having a separate test for it. It seems like it's just testing one more thing that appears on a row. The existing test is also checking the links, but we no longer have links on each of these.

muskangupta-iitr replied:
Yes, we can do that. I will update that.

bobsilverberg commented on tests/unit/amo/components/TestRatingsByStar.js:
```diff
@@ -218,25 +246,38 @@ describe(__filename, () => {
-      expect(link).toHaveProp(
-        'to',
-        reviewListURL({ addonSlug: addon.slug, score, src }),
+    const clientApp = CLIENT_APP_ANDROID;
```
See my comment above about setting these in the test.

muskangupta-iitr replied:
Yes! I updated this too.

bobsilverberg commented on tests/unit/amo/components/TestRatingsByStar.js:
```diff
@@ -218,25 +246,38 @@ describe(__filename, () => {
+    const lang = 'en-US';
+
+    function validateLink(row, score) {
+      expect(row).toHaveProp('onClick');
```
I don't think we need this assertion. We already have one test that checks that the `onClick` prop exists when expected, and the next assertion will fail if the `onClick` prop doesn't exist.

muskangupta-iitr replied:
I will remove this in the next update.

bobsilverberg commented on tests/unit/amo/components/TestRatingsByStar.js:
```diff
@@ -218,25 +246,38 @@ describe(__filename, () => {
 
   it('renders IconStar', () => {
     const root = render();
-    const star = root.find(IconStar).at(0);
```
Won't this still work? Are the changes to this test required?

muskangupta-iitr replied:
Yes, it is working fine without any changes. I will update that.

bobsilverberg commented on src/amo/components/RatingsByStar/index.js:
```diff
@@ -86,85 +91,100 @@ export class RatingsByStarBase extends React.Component<InternalProps> {
+
+      const rowProps =
+        loading || !addon
+          ? {}
```
I'm not sure we have a test for this logic. I.e., a test that asserts that `onClick` and `title` do not exist on the row when either `loading` is `true` or when we have no `addon`.

muskangupta-iitr replied:
I think we can check the `onClick` prop for row in test `renders a loading state without an add-on`. Is this correct?

bobsilverberg replied:
Yes, we could add those assertions to that test, but it looks like you haven't done that in the last commit.

muskangupta-iitr replied:
Sorry, just forgot to add this one. I will update this in the next commit.

bobsilverberg commented on src/amo/components/RatingsByStar/index.js:
```diff
@@ -86,85 +91,100 @@ export class RatingsByStarBase extends React.Component<InternalProps> {
+              <div className="RatingsByStar-bar RatingsByStar-barFrame" />
+            ) : (
+              <div className="RatingsByStar-bar RatingsByStar-barFrame">
+                {starCount !== undefined
```
Can you please check that we have a test for the case when `starCount` is `undefined`.

muskangupta-iitr replied:
No, we don't have any test for this case. I will add this in the update.

bobsilverberg replied:
I don't think that test is testing what I meant. If `starCount` is `undefined`, then rather than outputting the result of `this.renderBarValue(starCount)`, the code will output `null`, which would mean that there wouldn't even be a `div` with a `className` of `RatingsByStar-barValue`, but the test you mention (`renders 0% bar values when ratings do not exist`) is specifically checking elements with a `className` of `RatingsByStar-barValue`. 

So... either the code isn't working the way it is intended to, or we don't have a test where `starCount` is `undefined`.

muskangupta-iitr replied:
Yes, actually we don't have test when star count is undefined. I will add that in the next update.

muskangupta-iitr commented on src/amo/components/RatingsByStar/styles.scss:
```diff
@@ -1,24 +1,44 @@
+  border-spacing: 0 6px;
   color: $grey-50;
-  display: grid;
+  // display: grid;
```
Ohk! I will update that.

muskangupta-iitr commented:
@bobsilverberg I have added a commit updating the changes you suggested :+1:

muskangupta-iitr commented:
@bobsilverberg A gentle reminder in case you missed the comment :) I had completed the patch with all the suggested changes, could you please review it?

bobsilverberg commented:
My apologies @muskangupta-iitr. I was busy with some other things late last week, and then Friday was a holiday. I will take a look at this today.

bobsilverberg commented on tests/unit/amo/components/TestRatingsByStar.js:
```diff
@@ -146,61 +153,66 @@ describe(__filename, () => {
-      );
-      expect(link).toHaveProp('title', expectedTitle);
-    }
+    const customStore = dispatchClientMetadata({ clientApp, lang }).store;
```
You don't need to create a new `customStore` here. `dispatchClientMetadata` accepts a `store` argument, so you could just change this to `dispatchClientMetadata({ clientApp, lang, store })` and then use `store.dispatch` in the next line.

muskangupta-iitr replied:
Thanks! I will update that in next update.

bobsilverberg commented on tests/unit/amo/components/TestRatingsByStar.js:
```diff
@@ -213,30 +225,44 @@ describe(__filename, () => {
+    const clientApp = CLIENT_APP_FIREFOX;
+    const lang = 'en-US';
+
+    const customStore = dispatchClientMetadata({ clientApp, lang }).store;
```
As mentioned above, you don't need to create a new `store` to do this.

muskangupta-iitr commented:
> Thanks for the updates @muskangupta-iitr. I added some more comments about the tests.

Thanks @bobsilverberg for reviewing the updates. I will update the above changes in the next commit.

muskangupta-iitr commented:
@bobsilverberg I have updated the patch with the changes you suggested :+1:. Could you please review that?

bobsilverberg commented on tests/unit/amo/components/TestRatingsByStar.js:
```diff
@@ -123,7 +130,9 @@ describe(__filename, () => {
     const root = render({ addon: null });
+    const tableRow = root.find('.RatingByStar-table-row');
 
+    expect(tableRow.at(0)).not.toHaveProp('onClick');
```
Thanks for adding this. We should also have an assert that checks that the prop `title` also does not exist.

muskangupta-iitr replied:
Ohk! I will add that.

bobsilverberg commented on tests/unit/amo/components/TestRatingsByStar.js:
```diff
@@ -213,34 +227,65 @@ describe(__filename, () => {
+  it('renders when star count is undefined', () => {
+    let fiveStarCount;
+    const grouping = {
+      5: fiveStarCount,
```
Do you need to create this `fiveStarCount` variable, or could you just have `5: undefined,` here?

muskangupta-iitr replied:
Yes, we can directly write undefined but I just thought the case when `undefined` is possible, it happens only when variable is not initialized. But, yes it not necessary. I will update that.

bobsilverberg commented on tests/unit/amo/components/TestRatingsByStar.js:
```diff
@@ -213,34 +227,65 @@ describe(__filename, () => {
+    const root = render({ addon });
+    const bar = root.find('.RatingsByStar-bar');
+
+    expect(bar.at(0).children()).toHaveLength(0);
```
It would be nice to also have an assert for one of the bars that is not `undefined` which shows that `children()` has a length of `1`, so we can see that this test is in fact testing what we mean to test.

muskangupta-iitr replied:
Ohk! Nice, I will add this too!

muskangupta-iitr commented:
@bobsilverberg I have updated the changes and also did some styling changes as per the current design. Here are the screenshots:

![Screenshot from 2020-04-15 02-43-02](https://user-images.githubusercontent.com/50856599/79275179-58b1ef00-7ec3-11ea-9d2c-582727e43a96.png)

![Screenshot from 2020-04-15 02-43-06](https://user-images.githubusercontent.com/50856599/79275182-59e31c00-7ec3-11ea-9846-439da54c89d6.png)

bobsilverberg commented on src/amo/components/RatingsByStar/styles.scss:
```diff
@@ -2,23 +2,37 @@
-  // This controls whitespace and should also match the bar height.
   line-height: 16px;
+  width: 100%;
+  border-spacing: 0 8px;
```
This is going to fail our linter. Properties need to be specified in alphabetical order. You can also see this in the Travis results.

muskangupta-iitr replied:
Yes, Thanks for that @bobsilverberg. I noticed it now only. I will update that.

muskangupta-iitr commented:
> Looks great, thanks for all of your time and effort on this patch @muskangupta-iitr! There's just one small problem with the scss file that is causing the build to fail. Once that's fixed this should be good to land!

Thanks a lot @bobsilverberg for helping me completing this patch. I learned many things from this patch.

muskangupta-iitr commented:
@bobsilverberg Done!

muskangupta-iitr commented:
@bobsilverberg I have updated the title and description. Thanks for helping me out with the patch. Really glad to be able to contribute significantly. :)
